agent/util-scripts/pbench-tool-data-sink

```diff
@@ -0,0 +1,735 @@
+                    "Failed to rename tar ball '%s' to '%s'",
+                    ofp.name,
+                    host_tool_data_tb_md5,
+                )
+                raise
+            else:
+                self.logger.debug(
+                    "Successfully wrote %s (%s.md5)",
+                    host_tool_data_tb_name,
+                    host_tool_data_tb_name,
+                )
+
+        # Now unpack that tar ball
+        o_file = target_dir / f"{hostname}.tar.out"
+        e_file = target_dir / f"{hostname}.tar.err"
+        try:
+            # Invoke tar directly for efficiency.
+            with o_file.open("w") as ofp, e_file.open("w") as efp:
+                cp = subprocess.run(
+                    [tar_path, "-xf", host_tool_data_tb_name],
+                    cwd=target_dir,
+                    stdin=None,
+                    stdout=ofp,
+                    stderr=efp,
+                )
+        except Exception:
+            self.logger.exception(
+                "Failed to extract tools tar ball, '%s'", host_tool_data_tb_name
+            )
+            raise
+        else:
+            if cp.returncode != 0:
+                self.logger.error(
+                    "Failed to create tools tar ball; return code: %d", cp.returncode
+                )
+                abort(500, "INTERNAL ERROR")
+            else:
+                self.logger.debug("Successfully unpacked %s", host_tool_data_tb_name)
+                try:
+                    o_file.unlink()
+                    e_file.unlink()
+                    host_tool_data_tb_md5.unlink()
+                    host_tool_data_tb_name.unlink()
+                except Exception:
+                    self.logger.exception(
+                        "Error removing unpacked tar ball '%s' and it's .md5",
+                        host_tool_data_tb_name,
+                    )
+
+        # Tell the waiting "watcher" thread engaging in a "state" change
+        # that another PUT document has arrived.
+        with self._lock:
+            tm_tracker = self._tm_tracking[hostname]
+            assert tm_tracker["posted"] == "waiting", f"tm_tracker = {tm_tracker!r}"
+            tm_tracker["posted"] = "yes"
+            self._cv.notify()
+
+
+def main(argv):
+    PROG = Path(argv[0]).name
+
+    logger = logging.getLogger(PROG)
+    fh = logging.FileHandler(f"{PROG}.log")
+    if os.environ.get("_PBENCH_UNIT_TESTS"):
+        fmtstr = "%(levelname)s %(name)s %(funcName)s -- %(message)s"
+    else:
+        fmtstr = (
+            "%(asctime)s %(levelname)s %(process)s %(thread)s"
+            " %(name)s %(funcName)s %(lineno)d -- %(message)s"
+        )
+    fhf = logging.Formatter(fmtstr)
+    fh.setFormatter(fhf)
+    if os.environ.get("_PBENCH_TOOL_DATA_SINK_LOG_LEVEL") == "debug":
+        log_level = logging.DEBUG
+    else:
+        log_level = logging.INFO
+    fh.setLevel(log_level)
+    logger.addHandler(fh)
+    logger.setLevel(log_level)
```

Comment:
This basic setup is another thing that could be refactored into a common method somewhere, since there are several tools that do exact of very similar setup. But doesn't have to be now.

Reply:
Tracked via issue #1784.